python/openassetio_traitgen/generators/cpp.py

```diff
@@ -55,119 +61,190 @@ def generate(
+            imports.append(file_name)
+
+        # Generate the sub-package header that pre-imports all the
+        # namespaces for this "kind".
+        imports.sort()
+        docstring = f"{kind.capitalize()} defined in the '{self.__package.id}' package."
+        return self.__render_package_template(kind_abs_path, kind, docstring, imports)
+
+    def __render_namespace(
+        self, namespace: NamespaceDeclaration, parent_abs_path: str, kind: str
+    ) -> str:
+        namespace_module_name = self.__env.filters["to_cpp_module_name"](namespace.id)
+        namespace_abs_path = self.__create_dir_with_path_components(
+            parent_abs_path, namespace_module_name
+        )
+        imports = []
+
+        # Render a file per class (trait or specification).
+        for cls in namespace.members:
+            if kind == "traits":
+                file_name = self.__render_trait(namespace, cls, namespace_abs_path)
+            else:
+                file_name = self.__render_specification(namespace, cls, namespace_abs_path)
+
+            imports.append(f"{namespace_module_name}/{file_name}")
+
+        # Generate the namespace header that pre-imports all the
+        # classes.
+        imports.sort()
+        self.__render_template(
+            kind,
+            os.path.join(parent_abs_path, f"{namespace_module_name}.hpp"),
+            {
+                "package": self.__package,
+                "namespace": namespace,
+                "relImports": imports,
+                "traitgen_abi_version": TRAITGEN_ABI_VERSION,
+            },
+        )
+        return f"{namespace_module_name}.hpp"
+
+    def __render_trait(
+        self,
+        namespace: NamespaceDeclaration,
+        cls: TraitDeclaration,
+        namespace_abs_path: str,
+    ) -> str:
+        cls_name = self.__env.filters["to_cpp_class_name"](cls.name) + "Trait"
+        return self.__render_cls_template(namespace, cls, namespace_abs_path, "trait", cls_name)
+
+    def __render_specification(
+        self,
+        namespace: NamespaceDeclaration,
+        cls: SpecificationDeclaration,
+        namespace_abs_path: str,
+    ) -> str:
+        cls_name = self.__env.filters["to_cpp_class_name"](cls.id) + "Specification"
+        return self.__render_cls_template(
+            namespace, cls, namespace_abs_path, "specification", cls_name
+        )
+
+    def __render_package_template(
+        self, package_abs_path: str, name: str, docstring: str, imports: List[str]
+    ) -> str:
+        self.__render_template(
+            "package",
+            os.path.join(package_abs_path, f"{name}.hpp"),
+            {"docstring": docstring, "relImports": imports},
+        )
+        return f"{name}.hpp"
+
+    def __render_cls_template(
+        self,
+        namespace: NamespaceDeclaration,
+        cls: Union[SpecificationDeclaration, TraitDeclaration],
+        namespace_abs_path: str,
+        cls_kind: str,
```

Comment:
[fix] Can we disambiguate a bit? `cls` is passed to `cls_kind` internally, but it has its own `cls_kind` arg too?

Reply:
I'm not sure what you mean by "passed to `cls_kind` internally"?  Do you mean the template dict parameters using `cls_kind` as well as the `.__render_template` method?

Reply:
Yeah - was confusing having the name reused

Reply:
Refactored slightly in b4c1c14a - removed the common `__render_cls_template`  and renamed `cls` to `declaration`.